File: TidbitsTrivia/Core/Networking/WikipediaClient.swift
import Foundation

/// Read-only client for the open Wikipedia REST + Action APIs. No key, no
/// auth — so it deliberately bypasses APIClient's JWT path. `nonisolated`
/// so fetches run off the main actor (the app is MainActor by default).
nonisolated struct WikipediaClient: Sendable {
    static let shared = WikipediaClient()

    private let restBase = "https://en.wikipedia.org/api/rest_v1"
    private let actionBase = "https://en.wikipedia.org/w/api.php"

    // MARK: REST summary

    struct Summary: Decodable, Sendable {
        let title: String
        let description: String?
        let extract: String?
        let type: String?            // "standard" | "disambiguation" | ...
        let thumbnail: Thumb?
        let content_urls: URLs?
        struct Thumb: Decodable, Sendable { let source: String }
        struct URLs: Decodable, Sendable {
            let desktop: Page?
            struct Page: Decodable, Sendable { let page: String? }
        }
        var pageURL: URL? { content_urls?.desktop?.page.flatMap(URL.init(string:)) }
    }

    func summary(for title: String) async throws -> Summary {
        let path = title.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? title
        let url = URL(string: "\(restBase)/page/summary/\(path)")!
        var req = URLRequest(url: url)
        req.setValue("TidbitsTrivia/1.0 (learning trivia app)", forHTTPHeaderField: "User-Agent")
        let (data, resp) = try await URLSession.shared.data(for: req)
        guard let http = resp as? HTTPURLResponse, 200...299 ~= http.statusCode else {
            throw URLError(.badServerResponse)
        }
        return try JSONDecoder().decode(Summary.self, from: data)
    }

    // MARK: Search (Action API)

    struct SearchResult: Decodable, Sendable {
        let query: Query?
        struct Query: Decodable, Sendable { let search: [Hit] }
        struct Hit: Decodable, Sendable { let title: String }
    }

    /// Returns candidate article titles for a free-text topic.
    func search(_ topic: String, limit: Int = 30) async throws -> [String] {
        var comp = URLComponents(string: actionBase)!
        comp.queryItems = [
            .init(name: "action", value: "query"),
            .init(name: "list", value: "search"),
            .init(name: "srsearch", value: topic),
            .init(name: "srlimit", value: String(limit)),
            .init(name: "srnamespace", value: "0"),
            .init(name: "format", value: "json"),
        ]
        var req = URLRequest(url: comp.url!)
        req.setValue("TidbitsTrivia/1.0 (learning trivia app)", forHTTPHeaderField: "User-Agent")
        let (data, resp) = try await URLSession.shared.data(for: req)
        guard let http = resp as? HTTPURLResponse, 200...299 ~= http.statusCode else {
            throw URLError(.badServerResponse)
        }
        let decoded = try JSONDecoder().decode(SearchResult.self, from: data)
        return decoded.query?.search.map(\.title) ?? []
    }

    /// Fetch summaries for many titles concurrently, dropping failures.
    func summaries(for titles: [String]) async -> [Summary] {
        await withTaskGroup(of: Summary?.self) { group in
            for t in titles {
                group.addTask { try? await self.summary(for: t) }
            }
            var out: [Summary] = []
            for await s in group where s != nil { out.append(s!) }
            return out
        }
    }
}

File: windows/Tidbits.Core/Networking/WikipediaClient.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Http;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Threading.Tasks;

namespace Tidbits.Core.Networking;

/// Read-only client for the open Wikipedia REST + Action APIs (no key, no auth).
/// Port of Core/Networking/WikipediaClient.swift — feeds the TemplateEngine's
/// live question generation (1.11/1.12). The moat is the FILTER, not the fetch.
public sealed class WikipediaClient
{
    private const string RestBase = "https://en.wikipedia.org/api/rest_v1";
    private const string ActionBase = "https://en.wikipedia.org/w/api.php";
    private const string UA = "TidbitsTrivia/1.0 (learning trivia app)";

    private readonly HttpClient _http;
    public WikipediaClient(HttpClient? http = null)
    {
        _http = http ?? new HttpClient { Timeout = TimeSpan.FromSeconds(15) };
    }

    public sealed record Summary
    {
        [JsonPropertyName("title")] public string Title { get; init; } = "";
        [JsonPropertyName("description")] public string? Description { get; init; }
        [JsonPropertyName("extract")] public string? Extract { get; init; }
        [JsonPropertyName("type")] public string? Type { get; init; } // "standard" | "disambiguation" | …
        [JsonPropertyName("thumbnail")] public Thumb? Thumbnail { get; init; }
        [JsonPropertyName("content_urls")] public Urls? ContentUrls { get; init; }

        public sealed record Thumb { [JsonPropertyName("source")] public string? Source { get; init; } }
        public sealed record Urls
        {
            [JsonPropertyName("desktop")] public Page? Desktop { get; init; }
            public sealed record Page { [JsonPropertyName("page")] public string? PageUrl { get; init; } }
        }

        [JsonIgnore] public string? PageUrl => ContentUrls?.Desktop?.PageUrl;
        [JsonIgnore] public string? ImageUrl => Thumbnail?.Source;
    }

    /// Fetch a single article summary by exact title.
    public async Task<Summary?> GetSummary(string title)
    {
        var path = Uri.EscapeDataString(title.Replace(' ', '_'));
        using var req = new HttpRequestMessage(HttpMethod.Get, $"{RestBase}/page/summary/{path}");
        req.Headers.TryAddWithoutValidation("User-Agent", UA);
        using var resp = await _http.SendAsync(req);
        if (!resp.IsSuccessStatusCode) return null;
        var json = await resp.Content.ReadAsStringAsync();
        return Parse(json);
    }

    /// Parse a REST summary JSON body (extracted so it's unit-testable offline).
    public static Summary? Parse(string json)
    {
        try { return JsonSerializer.Deserialize<Summary>(json); }
        catch { return null; }
    }

    private sealed record SearchEnvelope
    {
        [JsonPropertyName("query")] public Query? Q { get; init; }
        public sealed record Query { [JsonPropertyName("search")] public List<Hit> Search { get; init; } = new(); }
        public sealed record Hit { [JsonPropertyName("title")] public string Title { get; init; } = ""; }
    }

    /// Candidate article titles for a free-text topic (Action API search).
    public async Task<List<string>> Search(string topic, int limit = 30)
    {
        var url = $"{ActionBase}?action=query&list=search&srsearch={Uri.EscapeDataString(topic)}" +
                  $"&srlimit={limit}&srnamespace=0&format=json";
        using var req = new HttpRequestMessage(HttpMethod.Get, url);
        req.Headers.TryAddWithoutValidation("User-Agent", UA);
        using var resp = await _http.SendAsync(req);
        if (!resp.IsSuccessStatusCode) return new();
        var json = await resp.Content.ReadAsStringAsync();
        return ParseSearch(json);
    }

    /// Parse an Action-API search body into titles (unit-testable offline).
    public static List<string> ParseSearch(string json)
    {
        try { return JsonSerializer.Deserialize<SearchEnvelope>(json)?.Q?.Search.Select(h => h.Title).ToList() ?? new(); }
        catch { return new(); }
    }

    /// Fetch summaries for many titles concurrently, dropping failures.
    public async Task<List<Summary>> Summaries(IEnumerable<string> titles)
    {
        var tasks = titles.Select(GetSummary).ToList();
        var results = await Task.WhenAll(tasks);
        return results.Where(s => s is not null).Select(s => s!).ToList();
    }
}
